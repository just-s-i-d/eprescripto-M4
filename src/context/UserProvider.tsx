import {
  PropsWithChildren,
  SetStateAction,
  createContext,
  useEffect,
  useState,
} from "react";

import { getUserData } from "@utils/Doctor";
import { showToast } from "@utils/common";

type UserContextType = {
  isLoggedIn?: boolean;
  setIsLoggedIn?: React.Dispatch<SetStateAction<boolean>>;
  isDarkTheme?: boolean;
  setIsDarkTheme?: React.Dispatch<SetStateAction<boolean>>;
  role?: string;
  setRole?: React.Dispatch<SetStateAction<string | undefined>>;
  refresh?: boolean;
  setRefresh?: React.Dispatch<SetStateAction<boolean | undefined>>;
  getRole?: () => void;
};

export const UserContext = createContext<UserContextType>({});
export const UserProvider = ({ children }: PropsWithChildren) => {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [isDarkTheme, setIsDarkTheme] = useState<boolean>(false);
  const [role, setRole] = useState<string>();

  const getRole = () => {
    getUserData()
      .then((res) => {
        setRole(res.role?.type);
      })
      .catch((error) => {
        showToast(error.message, error.type);
      });
  };
  useEffect(() => {
    const token = localStorage.getItem("token");
    if (token) {
      getRole();
    }
    setIsLoggedIn(!!token);
  }, []);
  const value = {
    isLoggedIn,
    setIsLoggedIn,
    role,
    setRole,
    isDarkTheme,
    setIsDarkTheme,
    getRole,
  };
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};
