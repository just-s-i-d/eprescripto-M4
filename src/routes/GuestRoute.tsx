import Loading from "@components/ui/Loading";
import { UserContext } from "@context/UserProvider";
import { useContext } from "react";
import { Navigate, Outlet } from "react-router-dom";

const GuestRoute = () => {
  const userContext = useContext(UserContext);
  if (!userContext) return <Loading />;
  const { isLoggedIn } = userContext;
  return (
    <>
      {!isLoggedIn && <Outlet />}
      {isLoggedIn && <Navigate to="/dashboard" />}
    </>
  );
};
export default GuestRoute;
