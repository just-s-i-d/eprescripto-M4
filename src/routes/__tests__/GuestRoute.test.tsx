import GuestRoute from "@routes/GuestRoute";
import "@testing-library/jest-dom";
import { screen, render } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { UserContext } from "@context/UserProvider";

type WrapperComponentProps = {
  loggedIn: boolean;
};
const WrapperComponent = ({ loggedIn }: WrapperComponentProps) => {
  return (
    <MemoryRouter initialEntries={["/auth"]}>
      <UserContext.Provider value={{ isLoggedIn: loggedIn }}>
        <Routes>
          <Route path="/auth" element={<GuestRoute />}>
            <Route index element={<h1>Guest Route</h1>} />
          </Route>
          <Route path="/dashboard" element={<h1>Dashboard</h1>} />
        </Routes>
      </UserContext.Provider>
    </MemoryRouter>
  );
};
describe("Guest Route tests", () => {
  it("Redirects to dashboard when user is logged in", () => {
    const { asFragment } = render(<WrapperComponent loggedIn={true} />);
    expect(screen.getByText("Dashboard")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Guest Route test", () => {
    render(<WrapperComponent loggedIn={false} />);
    expect(screen.getByText("Guest Route")).toBeInTheDocument();
  });
});
