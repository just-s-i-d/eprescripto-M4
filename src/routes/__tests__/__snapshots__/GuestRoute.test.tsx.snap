// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guest Route tests Redirects to dashboard when user is logged in 1`] = `
<DocumentFragment>
  <h1>
    Dashboard
  </h1>
</DocumentFragment>
`;
